### PR TITLE
o/snapstate: refactor Install functions to use a singular implementation that operates on a Target

### PR DESCRIPTION
**Pull request by andrewphelpsj**
The `Install` family of functions in the `snapstate` has grown, and there are a handful of functions now that increase our API's surface area.

Since a lot of these functions look really similar, we can change up the public API a bit to allow for a single `InstallTarget` function that operates on a `Target`, which controls which snaps are installed, and where they come from. In this PR, I've implemented this for installing snaps from the store and local snaps.

I was able to implement `Install`, `InstallMany`, `InstallPath`, `InstallWithDeviceContext`, and `InstallPathWithDeviceContext` using the new `InstallTarget` function. There are two target implementations in this PR, `StoreTarget` and `PathTarget`. `StoreTarget` allows the installation of one or many snaps from the store. `PathTarget` allows the installation of one local snap. Note that `PathTarget` does not support installing many local snaps at once, since that looks much more like an update, and should be included in a potential refactor of the `Update` family.

There is very little change to the current public API of the package, as I was able to implement all of the functions mentioned in terms of the new functions. The only minor change needed, outside of the functions themselves, are the changes inside of functions called from the `doPrerequisites` task handler. This handler manually set up some lanes for the installed snaps, since `InstallWithDeviceContext` did not support lanes. This change enables `InstallWithDeviceContext` to support lanes.

Also something that I took note of while I was in here that may not need to be addressed. We don't currently care about validation sets right now when installing local snaps. Should that happen? If so, I could move that code from the `StoreTarget` implementation to the `InstallTarget` function so that all installs always care about validation sets.

### CODE REVIEW

File: overlord/snapstate/target.go [hunk 1]
```diff
@@ -0,0 +1,536 @@
...
+			if !t.SkipIfPresent {
+				return &snap.AlreadyInstalledError{Snap: name}
+			}
+			delete(s.snaps, name)
+		}
+	}
+	return nil
+}
+
+func InstallOne(ctx context.Context, st *state.State, target Target, opts SnapstateOptions) (*snap.Info, *state.TaskSet, error) {
```
Comment (MiguelPires) [inline overlord/snapstate/target.go:286]: Comparing `InstallOne` with `InstallTarget` by their names, it's not clear which one does what. Especially since `Target` can be one or many snaps

File: overlord/snapstate/target.go [hunk 2]
```diff
@@ -0,0 +1,536 @@
...
+func validateRevisionOpts(opts RevisionOptions) error {
+	if opts.CohortKey != "" && opts.Revision.Set() {
+		return errors.New("cannot specify revision and cohort")
+	}
+
+	return nil
+}
+
+// Installables returns the data needed to setup the snaps from the store for
+// installation. Implements the Target interface.
```
Comment (MiguelPires) [inline overlord/snapstate/target.go:103]: nitpick
```suggestion
// installation.
```

File: overlord/snapstate/target.go [hunk 3]
```diff
@@ -0,0 +1,536 @@
...
+	opts.DeviceCtx, err = DevicePastSeeding(st, opts.DeviceCtx)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// PathTarget implements the Target interface and represents a single snap to be
+// installed from a path on disk.
```
Comment (MiguelPires) [inline overlord/snapstate/target.go:432]: same
```suggestion
// PathTarget represents a single snap to be installed from a path on disk.
```

File: overlord/snapstate/target.go [hunk 4]
```diff
@@ -0,0 +1,536 @@
...
+	opts.DeviceCtx, err = DeviceCtxFromState(st, opts.DeviceCtx)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Installables returns the data needed to setup the snap from disk. Implements
+// the Target interface.
```
Comment (MiguelPires) [inline overlord/snapstate/target.go:475]: same
```suggestion
// Installables returns the data needed to setup the snap from disk.
```

**Review by MiguelPires — COMMENTED** [review]
Thanks for this. I left a few comments but this looks much better

File: overlord/snapstate/target.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (andrewphelpsj) [reply overlord/snapstate/target.go:286]: I think the goal will be to replace it eventually. For example, when installing components (the motivator for this change), the old API won't get that functionality.

File: overlord/snapstate/target.go [hunk 5]
```diff
@@ -0,0 +1,536 @@
...
+	"github.com/snapcore/snapd/snap"
+	"github.com/snapcore/snapd/snap/naming"
+	"github.com/snapcore/snapd/store"
+)
+
+// SnapstateOptions contains optional parameters for the snapstate operations.
+// All of these fields are optional and can be left unset. The options in this
+// struct apply to all snaps that are part of an operation. Options that apply
+// to individual snaps can be found in RevisionOptions.
+type SnapstateOptions struct {
```
[the remark this answers is not in the document]
Comment (andrewphelpsj) [reply overlord/snapstate/target.go:20]: Good call, I had the same thought when I started using it to implement installing components from the store. Although, Samuele suggested that this type might not be something we want at all.

File: overlord/snapstate/target.go [hunk 6]
```diff
@@ -0,0 +1,536 @@
...
+		// one fails the changes for all affected snaps will be
+		// undone. Otherwise, have different lanes per snap so failures
+		// only affect the culprit snap.
+		switch opts.Flags.Transaction {
+		case client.TransactionAllSnaps:
+			return opts.Flags.Lane
+		case client.TransactionPerSnap:
+			return st.NewLane()
+		}
+		return 0
```
[the remark this answers is not in the document]
Comment (andrewphelpsj) [reply overlord/snapstate/target.go:354]: This isn't unexpected, sometimes the caller doesn't set anything for the `Transaction`. In that case, in the old API, we didn't mess with lanes at all.

File: overlord/snapstate/target.go [hunk 7]
```diff
@@ -0,0 +1,536 @@
...
+	if opts.CohortKey != "" && opts.Revision.Set() {
+		return errors.New("cannot specify revision and cohort")
+	}
+
+	return nil
+}
+
+// Installables returns the data needed to setup the snaps from the store for
+// installation. Implements the Target interface.
+func (s *StoreTarget) Installables(ctx context.Context, st *state.State, snaps map[string]*SnapState, opts SnapstateOptions) ([]Installable, error) {
```
[the remark this answers is not in the document]
Comment (andrewphelpsj) [reply overlord/snapstate/target.go:104]: Yes, that is a good idea.

File: overlord/snapstate/target.go (hunk 1 above)
Comment (andrewphelpsj) [reply overlord/snapstate/target.go:286]: Hrm, you're right. I'll think about names. In an ideal world, `InstallTarget` will become `Install` one day, once the old API stops being used.

File: overlord/snapstate/target.go [hunk 8]
```diff
@@ -0,0 +1,536 @@
...
+}
+
+// StoreTarget implements the TargetGroup interface and represents a group of
+// snaps that are to be installed from the store.
+type StoreTarget struct {
+	snaps map[string]StoreSnap
+}
+
+// verify that StoreTarget implements the Target interface
+var _ Target = &StoreTarget{}
```
[the remark this answers is not in the document]
Comment (andrewphelpsj) [reply overlord/snapstate/target.go:71]: We don't really need them, no. The main reason I added them in the first place was to verify that `&PathTarget{}` is an `OptionInitializer`, which wouldn't be compile-time checked anywhere else.

File: overlord/snapstate/target.go [hunk 9]
```diff
@@ -0,0 +1,536 @@
...
+	// can only specify a lane when running multiple operations transactionally
+	if opts.Flags.Transaction != client.TransactionAllSnaps && opts.Flags.Lane != 0 {
+		return nil, nil, errors.New("cannot specify a lane without setting transaction to \"all-snaps\"")
+	}
+
+	if opts.Flags.Transaction == client.TransactionAllSnaps && opts.Flags.Lane == 0 {
+		opts.Flags.Lane = st.NewLane()
+	}
+
+	if initer, ok := target.(OptionInitializer); ok {
```
[the remark this answers is not in the document]
Comment (andrewphelpsj) [reply overlord/snapstate/target.go:312]: That is a good thought, I'll try implementing it that way too.

File: overlord/snapstate/target.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (andrewphelpsj) [reply overlord/snapstate/target.go:103]: I would also love to get rid of that parameter completely, tbh. But one of the our tests shows some behavior that indicates that we need to get the list of installed snaps prior to unlocking the state lock, if we happen to.

Basically, we need to compare the state of the installed snaps before we unlock the lock (to talk to the store) to after we unlock the lock, to make sure nothing happened in between: https://github.com/snapcore/snapd/blob/master/overlord/snapstate/snapstate_install_test.go?plain=1#L986

File: overlord/snapstate/handlers_prereq_test.go [hunk 10]
```diff
@@ -680,7 +680,7 @@ func (s *prereqSuite) TestDoPrereqBaseNoRevision(c *C) {
 	s.state.Lock()
 	defer s.state.Unlock()
 	c.Check(chg.Status(), Equals, state.ErrorStatus)
-	c.Check(chg.Err(), ErrorMatches, `cannot perform the following tasks:\n.*- test \(cannot install snap base "some-base": no snap revision available as specified\)`)
+	c.Check(chg.Err(), ErrorMatches, `cannot perform the following tasks:\n.*- test \(cannot install snap base "some-base": cannot install snap "some-base": no snap revision available as specified\)`)
```
Comment (pedronis) [inline overlord/snapstate/handlers_prereq_test.go:683]: this change to errors is weird with the repetition, it would be good to work to avoid it somehow

File: overlord/snapstate/handlers.go [hunk 11]
```diff
@@ -476,6 +481,21 @@ func hasAllContentAttrs(st *state.State, snapName string, requiredContentAttrs [
...
+	if flags.Transaction == client.TransactionAllSnaps {
+		lanes := t.Lanes()
+		if len(lanes) != 1 {
+			return fmt.Errorf("internal error: more than one lane (%d) on a transactional action", len(lanes))
+		}
+
+		flags.Lane = lanes[0]
+	} else {
+		flags.Transaction = client.TransactionPerSnap
+	}
```
Comment (pedronis) [inline overlord/snapstate/handlers.go:497]: this logic replaces the one below using joinLane?

File: overlord/snapstate/target.go [hunk 12]
```diff
@@ -0,0 +1,552 @@
...
+type OptionInitializer interface {
+	// InitOptions initializes the SnapstateOptions before the installation of
+	// the snaps.
+	InitOptions(*state.State, *Options) error
+}
+
+// StoreSnap represents a snap that is to be installed from the store.
+type StoreSnap struct {
+	// Name is the name of snap to install.
+	Name string
```
Comment (pedronis) [inline overlord/snapstate/target.go:74]: is this a snap name or an instance name? it should be clearer

File: overlord/snapstate/target.go [hunk 13]
```diff
@@ -0,0 +1,552 @@
...
+// Options contains optional parameters for the snapstate operations. All of
+// these fields are optional and can be left unset. The options in this struct
+// apply to all snaps that are part of an operation. Options that apply to
+// individual snaps can be found in RevisionOptions.
+type Options struct {
+	Flags         Flags
+	UserID        int
+	FromChange    string
+	PrereqTracker PrereqTracker
+	DeviceCtx     DeviceContext
```
Comment (pedronis) [inline overlord/snapstate/target.go:44]: the order of the fields should be:

DeviceCtx, PrereqTracker, FromChange

Comment (pedronis) [reply overlord/snapstate/target.go:44]: fields need a doc comment

File: overlord/snapstate/target.go [hunk 14]
```diff
@@ -0,0 +1,552 @@
...
+}
+
+// Installable represents the data needed to setup a snap for installation.
+type Installable struct {
+	// Snap indicates how to find the snap to install.
```
Comment (pedronis) [inline overlord/snapstate/target.go:49]: Snap describes the snap to install ?

File: overlord/snapstate/target.go [hunk 15]
```diff
@@ -0,0 +1,552 @@
...
+	Snap *SnapSetup
+	// Components indicates how to find the components to install.
```
Comment (pedronis) [inline overlord/snapstate/target.go:51]: same with describe instead of find?

File: overlord/snapstate/target.go [hunk 16]
```diff
@@ -0,0 +1,552 @@
...
+
+// Target represents a single snap or a group of snaps to be installed.
+type Target interface {
+	// Installables returns the data needed to setup the snaps for installation.
+	Installables(context.Context, *state.State, map[string]*SnapState, Options) ([]Installable, error)
+}
+
+// OptionInitializer is an interface that can be implemented by a Target to
+// initialize the SnapstateOptions before the installation of the snaps.
+type OptionInitializer interface {
```
Comment (pedronis) [inline overlord/snapstate/target.go:65]: does this interface need to be public?

File: overlord/snapstate/target.go [hunk 17]
```diff
@@ -0,0 +1,552 @@
...
+type Options struct {
+	Flags         Flags
+	UserID        int
+	FromChange    string
+	PrereqTracker PrereqTracker
+	DeviceCtx     DeviceContext
+}
+
+// Installable represents the data needed to setup a snap for installation.
+type Installable struct {
```
Comment (pedronis) [inline overlord/snapstate/target.go:48]: we need a better name for this, it's strange to use an adjective for a struct

File: overlord/snapstate/target.go [hunk 18]
```diff
@@ -0,0 +1,552 @@
...
+		InstanceName: name,
+		Path:         path,
+		RevOpts:      opts,
+		SideInfo:     si,
+	}
+}
+
+// InitOptions initializes the SnapstateOptions before the installation of the
+// snaps. Implements the OptionInitializer interface.
+func (p *PathTarget) InitOptions(st *state.State, opts *Options) error {
```
Comment (pedronis) [inline overlord/snapstate/target.go:481]: it's strange for this to be public

File: overlord/snapstate/target.go [hunk 19]
```diff
@@ -0,0 +1,552 @@
...
+		return opts.Flags.Lane
+	case client.TransactionPerSnap:
+		return st.NewLane()
+	}
+	return 0
+}
+
+func setDefaultSnapstateOptions(st *state.State, opts *Options) error {
+	var err error
+	opts.DeviceCtx, err = DevicePastSeeding(st, opts.DeviceCtx)
```
Comment (pedronis) [inline overlord/snapstate/target.go:441]: I also think we probably want to express this difference differently now that we have Options, I think the issue is more first boot code vs later code

File: overlord/snapstate/target.go [hunk 20]
```diff
@@ -0,0 +1,552 @@
...
+			if !t.SkipIfPresent {
+				return &snap.AlreadyInstalledError{Snap: name}
+			}
+			delete(s.snaps, name)
+		}
+	}
+	return nil
+}
+
+func InstallOne(ctx context.Context, st *state.State, target Target, opts Options) (*snap.Info, *state.TaskSet, error) {
```
Comment (pedronis) [inline overlord/snapstate/target.go:304]: this needs a doc comment

File: overlord/snapstate/target.go [hunk 21]
```diff
@@ -0,0 +1,552 @@
...
+	}
+
+	if len(infos) != 1 || len(tasksets) != 1 {
+		return nil, nil, errors.New("internal error: expected exactly one snap and taskset")
+	}
+
+	return infos[0], tasksets[0], nil
+}
+
+func InstallTarget(ctx context.Context, st *state.State, target Target, opts Options) ([]*snap.Info, []*state.TaskSet, error) {
```
Comment (pedronis) [inline overlord/snapstate/target.go:317]: same

File: overlord/snapstate/target.go [hunk 22]
```diff
@@ -0,0 +1,552 @@
...
+	}
+
+	return &StoreTarget{
+		snaps: mapping,
+	}
+}
+
+func validateRevisionOpts(opts RevisionOptions) error {
+	if opts.CohortKey != "" && opts.Revision.Set() {
+		return errors.New("cannot specify revision and cohort")
```
Comment (pedronis) [inline overlord/snapstate/target.go:114]: this doesn't seem to be exercised by tests

File: overlord/snapstate/target.go [hunk 23]
```diff
@@ -0,0 +1,552 @@
...
+	}
+
+	return nil
+}
+
+// Installables returns the data needed to setup the snaps from the store for
+// installation.
+func (s *StoreTarget) Installables(ctx context.Context, st *state.State, installedSnaps map[string]*SnapState, opts Options) ([]Installable, error) {
```
Comment (pedronis) [inline overlord/snapstate/target.go:122]: this is mostly replacing installCandidates?

File: overlord/snapstate/target.go [hunk 24]
```diff
@@ -0,0 +1,552 @@
...
+		ts.JoinLane(laneGenerator(st, opts))
+
+		tasksets = append(tasksets, ts)
+		infos = append(infos, info)
+	}
+
+	return infos, tasksets, nil
+}
+
+func laneGenerator(st *state.State, opts Options) int {
```
Comment (pedronis) [inline overlord/snapstate/target.go:425]: generateLane?

File: overlord/snapstate/target.go [hunk 25]
```diff
@@ -0,0 +1,552 @@
...
+		}
+
+		if err := validateRevisionOpts(t.RevOpts); err != nil {
+			return fmt.Errorf("invalid revision options for snap %q: %w", name, err)
+		}
+
+		snapst, ok := installedSnaps[name]
+		if ok && snapst.IsInstalled() {
+			if !t.SkipIfPresent {
+				return &snap.AlreadyInstalledError{Snap: name}
```
Comment (pedronis) [inline overlord/snapstate/target.go:296]: we don't seem to arrive here in tests

File: overlord/hookstate/ctlcmd/services_test.go [hunk 26]
```diff
@@ -222,7 +222,7 @@ func (s *servicectlSuite) SetUpTest(c *C) {
 		snapstate.EnforcedValidationSets = old
 	})
 	snapstate.EnforcedValidationSets = func(st *state.State, extraVss ...*asserts.ValidationSet) (*snapasserts.ValidationSets, error) {
-		return nil, nil
+		return snapasserts.NewValidationSets(), nil
```
Comment (pedronis) [inline overlord/hookstate/ctlcmd/services_test.go:225]: what's the reason for the change?

**Review by pedronis — DISMISSED** [review]
thank you, did a pass, direction looks good, there is some cleaning up to do though

File: overlord/hookstate/ctlcmd/services_test.go (hunk 26 above)
Comment (andrewphelpsj) [reply overlord/hookstate/ctlcmd/services_test.go:225]: In the previous version of the code that checked the validation sets against the to-be-installed snaps, we had `nil` checks in places where we had already confirmed that the error returned from `snapstate.EnforcedValidationSets` was `nil`. Since the actual implementation of `snapstate.EnforcedValidationSets` never returns `nil, nil`, I was able to simplify the calling code to eliminate those redundant `nil` checks, but I had to update the mocked implementations.

File: overlord/snapstate/handlers.go (hunk 11 above)
Comment (andrewphelpsj) [reply overlord/snapstate/handlers.go:497]: Correct, now that `InstallWithDeviceContext` supports lanes, we can use that rather than manually doing it here.

File: overlord/snapstate/target.go (hunk 19 above)
Comment (andrewphelpsj) [reply overlord/snapstate/target.go:447]: Are you thinking we should add some field to Options that indicates that the system doesn't need to be seeded? Then it could be used in the first boot code?

File: overlord/snapstate/target.go [hunk 27]
```diff
@@ -0,0 +1,569 @@
...
+	// SnapState is the current state of the target snap, prior to installation.
+	// This must be retrieved prior to unlocking the state for any reason (for
+	// example, talking to the store).
+	SnapState SnapState
+}
+
+// InstallGoal represents a single snap or a group of snaps to be installed.
+type InstallGoal interface {
+	// Install returns the data needed to setup the snaps for installation.
+	Install(context.Context, *state.State, Options) ([]Target, error)
```
Comment (pedronis) [inline overlord/snapstate/target.go:79]: shouldn't this be ToInstall or Targets ?

File: overlord/snapstate/target.go [hunk 28]
```diff
@@ -0,0 +1,569 @@
...
+// an error is returned.
+func InstallOne(ctx context.Context, st *state.State, goal InstallGoal, opts Options) (*snap.Info, *state.TaskSet, error) {
+	infos, tasksets, err := InstallWithGoal(ctx, st, goal, opts)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	if len(infos) != 1 || len(tasksets) != 1 {
+		return nil, nil, errors.New("internal error: expected exactly one snap and taskset")
+	}
```
Comment (pedronis) [inline overlord/snapstate/target.go:344]: this is a bit strange because one would expect that we could check whether goal return one target or not earlier

File: overlord/snapstate/target.go [hunk 29]
```diff
@@ -0,0 +1,569 @@
...
+	// FromChange is the change that triggered the operation.
+	FromChange string
+	// Seed should be true while seeding the device. This indicates that we
+	// shouldn't require that the device is seeded before installing/updating
+	// snaps.
+	Seed bool
+}
+
+// Target represents the data needed to setup a snap for installation.
+type Target struct {
```
Comment (pedronis) [inline overlord/snapstate/target.go:60]: as we discussed, should we make more things unexported ?

File: overlord/snapstate/target.go [hunk 30]
```diff
@@ -0,0 +1,569 @@
...
+	}
+
+	str := Store(st, opts.DeviceCtx)
+
+	st.Unlock() // calls to the store should be done without holding the state lock
+	defer st.Lock()
+
+	results, _, err := str.SnapAction(context.TODO(), curSnaps, actions, nil, user, refreshOpts)
+	if err != nil {
+		if len(actions) == 1 {
```
Comment (pedronis) [inline overlord/snapstate/target.go:193]: I'm not sure this is quite the right approach, passing 1 snap to a Many function previously has different error behavior than passing it to a single target function. We might need to have a Single flag in options or something to control this better

**Review by pedronis — COMMENTED** [review]
thanks for the changes, did another quick pass with some comments on target.go

File: overlord/snapstate/target.go (hunk 29 above)
Comment (andrewphelpsj) [reply overlord/snapstate/target.go:60]: If this is unexported, then the interface should be unexported as well.

File: overlord/snapstate/target.go (hunk 28 above)
Comment (andrewphelpsj) [reply overlord/snapstate/target.go:344]: Yes, this is a bit annoying. I was trying to avoid a flag being passed into `InstallWithGoal` to check that. One option is some implementation of `InstallGoal` that wraps another `InstallGoal` that makes sure that we return a singular `Target`.

Comment (andrewphelpsj) [reply overlord/snapstate/target.go:344]: Might be a bit too complex though for something that really could just be an `if`.

File: overlord/snapstate/target.go (hunk 30 above)
Comment (andrewphelpsj) [reply overlord/snapstate/target.go]: Added in 938cd26c6694412ea13e8bb443e0c164b3aa6e1f.

File: overlord/snapstate/target.go (hunk 29 above)
Comment (andrewphelpsj) [reply overlord/snapstate/target.go]: Done in bda59b73215fa73c3c64f062f8a73217d6b66052.

File: overlord/snapstate/target.go (hunk 28 above)
Comment (andrewphelpsj) [reply overlord/snapstate/target.go]: This can also be handled with the RequireOneSnap option introduced in 938cd26c6694412ea13e8bb443e0c164b3aa6e1f.

**Comment by pedronis** [comment]
@andrewphelpsj there is a conflict now

File: overlord/snapstate/target.go [hunk 31]
```diff
@@ -63,30 +63,30 @@ type Options struct {
...
-	SnapState SnapState
+	snapst SnapState
 }
 
-// InstallGoal represents a single snap or a group of snaps to be installed.
-type InstallGoal interface {
-	// ToInstall returns the data needed to setup the snaps for installation.
-	ToInstall(context.Context, *state.State, Options) ([]Target, error)
+// installGoal represents a single snap or a group of snaps to be installed.
+type installGoal interface {
```
Comment (pedronis) [inline overlord/snapstate/target.go:84]: this one should probably stay exported, is usually strange to have input to an exported  function that are unexported

File: overlord/snapstate/target.go [hunk 32]
```diff
@@ -104,10 +104,10 @@ type StoreSnap struct {
...
 
-// verify that StoreTarget implements the Target interface
-var _ InstallGoal = &storeInstallGoal{}
+// verify that storeInstallGoal implements the installGoal interface
+var _ installGoal = &storeInstallGoal{}
 
-// StoreGoal creates a new storeInstallGoal, which implements InstallGoal to
+// StoreGoal creates a new storeInstallGoal, which implements installGoal to
 // install snaps from the store.
 func StoreGoal(snaps ...StoreSnap) *storeInstallGoal {
```
Comment (pedronis) [inline overlord/snapstate/target.go:112]: this should use the interface as return type? same for the other Goal constructor

**Review by pedronis — COMMENTED** [review]
thanks for the changes, couple quick comments

File: overlord/snapstate/target.go [hunk 33]
```diff
@@ -0,0 +1,588 @@
...
+	// shouldn't require that the device is seeded before installing/updating
+	// snaps.
+	Seed bool
+	// RequireOneSnap is a boolean flag indicating that this operation is expected
+	// to only operate on one snap (excluding any prerequisite snaps that may be
+	// required). If this is true, then the operation will fail if more than one
+	// snap is being operated on. This flag primarily exists to support the
+	// pre-existing behavior of calling InstallMany with one snap vs calling
+	// Install.
+	RequireOneSnap bool
```
Comment (pedronis) [inline overlord/snapstate/target.go:63]: wouldn't ExpectOneSnap be clearer for this? OTOH I would also be ok with calling this SingleSnap

File: overlord/snapstate/target.go [hunk 34]
```diff
@@ -0,0 +1,588 @@
...
+	InstanceName string
+	// Components is the list of components to install with this snap.
+	Components []string
+	// RevOpts contains options that apply to the installation of this snap.
+	RevOpts RevisionOptions
+	// SkipIfPresent indicates that the snap should not be installed if it is already present.
+	SkipIfPresent bool
+}
+
+// StoreInstallGoal creates a new storeInstallGoal, which implements InstallGoal to
```
Comment (pedronis) [inline overlord/snapstate/target.go:107]: we want to rephrase this leaving out the internal name, something like:

creates/returns an implementation of InstallGoal to install/installing ...

File: overlord/snapstate/target.go [hunk 35]
```diff
@@ -0,0 +1,588 @@
...
+
+func validateRevisionOpts(opts RevisionOptions) error {
+	if opts.CohortKey != "" && opts.Revision.Set() {
+		return errors.New("cannot specify revision and cohort")
+	}
+
+	return nil
+}
+
+var ErrExpectedOneSnap = errors.New("expected exactly one snap to install")
```
Comment (pedronis) [inline overlord/snapstate/target.go:136]: given this Expect in the option name would make more sense

File: overlord/snapstate/target.go [hunk 36]
```diff
@@ -0,0 +1,588 @@
...
+	// instanceName is the name of the snap to install.
+	instanceName string
+	// revOpts contains options that apply to the installation of this snap.
+	revOpts RevisionOptions
+	// sideInfo contains extra information about the snap.
+	sideInfo *snap.SideInfo
+}
+
+// PathInstallGoal creates a new pathInstallGoal from the given instance name,
+// path, and side info.
```
Comment (pedronis) [inline overlord/snapstate/target.go:511]: same comment about mentioning the internal struct

File: overlord/snapstate/target.go [hunk 37]
```diff
@@ -0,0 +1,588 @@
...
+func PathInstallGoal(instanceName, path string, si *snap.SideInfo, opts RevisionOptions) InstallGoal {
```
Comment (pedronis) [inline overlord/snapstate/target.go:512]: the doc comment should explain that if instanceName is ""  si.RealName will be used

File: snap/revision.go [hunk 38]
```diff
@@ -44,6 +44,10 @@ func (r Revision) Unset() bool {
 	return r.N == 0
 }
 
+func (r Revision) Set() bool {
```
Comment (pedronis) [inline snap/revision.go:47]: as we have already quite a bit of code using !Unset, I'm not entirely sure if this addition is worth it at this time

File: overlord/devicestate/firstboot.go [hunk 39]
```diff
@@ -54,7 +55,17 @@ func installSeedSnap(st *state.State, sn *seed.Snap, flags snapstate.Flags, prqt
 		flags.DevMode = true
 	}
 
-	return snapstate.InstallPath(st, sn.SideInfo, sn.Path, "", sn.Channel, flags, prqt)
+	t := snapstate.PathInstallGoal(sn.SideInfo.RealName, sn.Path, sn.SideInfo, snapstate.RevisionOptions{})
```
Comment (pedronis) [inline overlord/devicestate/firstboot.go:58]: sn.SideInfo.RealName could just be "" afaiu

**Review by pedronis — COMMENTED** [review]
thank you, did another pass, some final minor things

File: overlord/snapstate/target.go [hunk 40]
```diff
@@ -0,0 +1,587 @@
...
+	// instanceName is the name of the snap to install.
+	instanceName string
+	// revOpts contains options that apply to the installation of this snap.
+	revOpts RevisionOptions
+	// sideInfo contains extra information about the snap.
+	sideInfo *snap.SideInfo
+}
+
+// PathInstallGoal creates a new InstallGoal to install a snap from a given from
+// a path on disk. If instanceName is not provided, si.RealName will be used.
```
Comment (pedronis) [inline overlord/snapstate/target.go:510]: would "empty" be clearer than "is not provided"?

**Review by pedronis — APPROVED** [review]
thank you

File: overlord/snapstate/target.go [hunk 41]
```diff
@@ -0,0 +1,587 @@
...
+
+// target represents the data needed to setup a snap for installation.
+type target struct {
+	// setup is a partially initialized SnapSetup that contains the data needed
+	// to find the snap file that will be installed.
+	setup *SnapSetup
+	// components is a list of partially initialized ComponentSetup structs that
+	// contain the data needed to find the component files that will be
+	// installed.
+	components []*ComponentSetup
```
Comment (MiguelPires) [inline overlord/snapstate/target.go:74]: I see there's a TODO below for this, but golangci complains so maybe we can comment this out and leave a similar TODO

**Review by MiguelPires — APPROVED** [review]
Thank you, LGTM. Just some minor questions

File: overlord/snapstate/target.go [hunk 42]
```diff
@@ -0,0 +1,587 @@
...
+// snaps from. The Options struct contains optional parameters that apply to the
+// installation operation.
+//
+// A slice of snap.Info structs is returned for each snap that is being
+// installed along with a slice of state.TaskSet structs that represent the
+// tasks that are part of the installation operation for each snap.
+//
+// TODO: rename this to Install once the API is settled, and we can rename or
+// remove the old Install function.
+func InstallWithGoal(ctx context.Context, st *state.State, goal InstallGoal, opts Options) ([]*snap.Info, []*state.TaskSet, error) {
```
[the remark this answers is not in the document]
Comment (pedronis) [reply overlord/snapstate/target.go:376]: I discussed with @andrewphelpsj to create an install.go file as well, maybe it could an immediate follow-up

File: overlord/snapstate/target.go [hunk 43]
```diff
@@ -0,0 +1,587 @@
...
+		}
+
+		opts.PrereqTracker.Add(info)
+
+		flags, err := earlyChecks(st, &inst.snapst, info, opts.Flags)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		providerContentAttrs := defaultProviderContentAttrs(st, info, opts.PrereqTracker)
```
[the remark this answers is not in the document]
Comment (andrewphelpsj) [reply overlord/snapstate/target.go:434]: I do agree that I don't love passing around the partially initialized `SnapSetup`. I think maybe another embedded goal might be too overkill, but it should be easy to just abstract finishing the `SnapSetup` into a function that could be shared by the `InstallGoal` implementations? Thoughts on that?

[the remark this answers is not in the document]
Comment (pedronis) [reply overlord/snapstate/target.go:434]: can't it not be a method on Target?

[the remark this answers is not in the document]
Comment (andrewphelpsj) [reply overlord/snapstate/target.go]: That is a good idea, and it will be helpful in the update implementation as well.

[the remark this answers is not in the document]
Comment (andrewphelpsj) [reply overlord/snapstate/target.go]: @MiguelPires changed in ce46744d045417a11297ada33eac124fe7ecec33

File: overlord/snapstate/target.go (hunk 42 above)
[the remark this answers is not in the document]
Comment (andrewphelpsj) [reply overlord/snapstate/target.go:409]: Yeah, I'll move things around in a follow-up.

**Review by alfonsosanchezbeato — APPROVED** [review]
Thanks, this is a nice clean-up. I have a couple of minor comments, but LGTM otherwise.

File: overlord/snapstate/target.go [hunk 44]
```diff
@@ -0,0 +1,599 @@
...
+	SkipIfPresent bool
+}
+
+// StoreInstallGoal creates a new InstallGoal to install snaps from the store.
+func StoreInstallGoal(snaps ...StoreSnap) InstallGoal {
+	mapping := make(map[string]StoreSnap, len(snaps))
+	for _, sn := range snaps {
+		if _, ok := mapping[sn.InstanceName]; ok {
+			continue
+		}
```
[the remark this answers is not in the document]
Comment (andrewphelpsj) [reply overlord/snapstate/target.go]: I made that change and clarified the behavior in the doc comment.
